ValkeyLDAP.md

```diff
@@ -0,0 +1,271 @@
+- Connection failover
+- Cache replication
+- Graceful degradation
+
+### Error Handling
+
+- Invalid credentials
+- Network timeouts
+- LDAP server unavailable
+- Cache corruption
+- Configuration errors
+- Group resolution failures
+
+## Specification
+
+### Authentication flow
+
+1. Client sends AUTH command
+2. Module extracts username/password
+3. Check credential cache
+4. If not cached:
+    - connect to LDAP server
+    - perform bind operation
+    - search for user group membership
+    - cache successful result
+    - set TTL
+5. Pass AUTH/NOAUTH to Valkey core
+6. Return auth response
+
+### On module load
+
+- Load configuration from Valkey config or default values
+- Initialize data structures and connection pools
+- Register command hooks with valkey core
+- Establish initial LDAP server connections
+- Load cached credentials if persistence enabled
+- Register monitoring callbacks
+
+### Commands 
+
+1 LDAP.CONFIG SET <parameter> <value>
+
+- Sets configuration parameters for the LDAP authentication module
+- Parameters include server addresses, timeouts, cache settings, etc.
+- Returns simple string reply: OK on success
+
+2 LDAP.CONFIG GET <parameter>
+
+- Gets current configuration for specified parameter
+- Returns bulk string reply: parameter value
+
+3 LDAP.RELOAD
+
+- Reloads configuration from config file
+- Reestablishes connections with LDAP servers
+- Returns simple string reply: OK on success
+
+4 LDAP.STATUS
+
+- Returns health and statistics information
+- Format: nested hash with connection states, cache stats, auth counts
+
+5 LDAP.FLUSHCACHE
+
+- Clears the credential cache
+- Options to flush all or specific users
+- Returns integer reply: number of entries flushed
+
+### Valkey Command Hooks
+
+#### Authentication Command Hook
+
+- Intercepts AUTH command
+- Processes credentials through LDAP binding
+- Maintains original Valkey auth functionality for local accounts
+
+#### ACL Command Hook
+
+- Integrates with ACL commands
+- Maps LDAP groups to Valkey ACLs
+- Preserves ACL inheritance and evaluation logic
+
+### Internal data structures
+
+#### Configuration Store
+
+- Hash table for configuration parameters
+- Persistent across restarts via RDB or config file
+- Thread-safe access with read-write locks
+
+#### Credential Cache
+
+- LRU cache with configurable TTL
+- Encrypted storage with secure key management
+- Size-limited with automatic eviction
+- Concurrent access optimized
+
+#### Group Mapping Table
+
+- Hash table mapping LDAP groups to Valkey ACLs
+- Configurable refresh interval
+- Support for nested group resolution
+
+#### Connection Pool
+
+- LDAP connection objects with state tracking
+- Priority queue based on connection health
+- Automatic connection cycling
+- Size limits and idle timeouts
+
+### Configuration
+
+#### Server Configuration
+
+- LDAP.servers: list of LDAP server addresses (comma-separated)
+- LDAP.port: LDAP port (default: 636 for LDAPS)
```

Comment:
Should this just be a list of LDAP servers + ports. It could also include the TLS information. Reminds me of how it was brought up in the past that having a map or nested structure would be nice for configs.